pytorch/imagerecognition/openmpi-cifar10-resnet20-all-reduce-scaling/main.py

```diff
@@ -169,19 +169,30 @@ def train_loop(run_id, dataset_dir, ckpt_run_dir, output_dir,
             json.dump(val_stats, f)
 
 
-def main(run_id, dataset_dir, ckpt_run_dir, output_dir, validation_only=False,
+def main(run_id, dataset_dir, ckpt_run_dir, output_dir, rank, world_size, backend, hosts, validation_only=False,
          gpu=False, light_target=False):
     r"""Main logic."""
-
-    with initialize_backends(
-            comm_backend='mpi',
-            logging_level='INFO',
-            logging_file=os.path.join(output_dir, 'mlbench.log'),
-            use_cuda=gpu,
-            seed=42,
-            cudnn_deterministic=False,
-            ckpt_run_dir=ckpt_run_dir,
-            delete_existing_ckpts=not validation_only):
+    if backend == "mpi":
+        with initialize_backends(
+                comm_backend='mpi',
+                logging_level='INFO',
+                logging_file=os.path.join(output_dir, 'mlbench.log'),
+                use_cuda=gpu,
+                seed=42,
+                cudnn_deterministic=False,
+                ckpt_run_dir=ckpt_run_dir,
+                delete_existing_ckpts=not validation_only):
+            train_loop(run_id, dataset_dir, ckpt_run_dir, output_dir,
+                       validation_only, use_cuda=gpu, light_target=light_target)
+    elif backend == "gloo":
+        hosts = hosts.split(',')
+        os.environ['MASTER_ADDR'] = hosts[0]  # first worker is the master worker
+        os.environ['MASTER_PORT'] = '29500'
```

Comment:
Does this work? Usually you need also expose ports in the Dockerfile and kubernetes also needs the ports open. 

Currently, only ssh is opened by default, see https://github.com/mlbench/mlbench-dashboard/blob/develop/src/api/utils/run_utils.py#L107 local port 22 is exposed as host port 16166. This has the side effect to force one pod per node, though this is also handled by k8s affinities. Either we'd have to reuse this port (with ssh disabled) or add some mechanism/config that allows exposing additional ports in k8s. Unless of course this already works for some reason (which it shouldn't as far as I understand k8s)

Reply:
I tried it and the training was working fine. I did not wait until the finish to avoid unnecessary expenses, but I started getting partial results, so I think there is no problem. I can test with a full run, but I think that if there was a problem it would not have started at all. What do you think?